### ISSUE TITLE
Package request: uftrace

### ISSUE THREAD

**Issue by twaik** [termux/termux-packages#4216]
**Package description**
The uftrace tool is to trace and analyze execution of a program written in C/C++

**Link to home page and sources**
1. Home page: https://uftrace.github.io/slide/
2. Source code: https://github.com/namhyung/uftrace

**Additional information**
NDK misses some functions, does not build. 
[uftrace.zip](https://github.com/termux/termux-packages/files/3537616/uftrace.zip)

It will let us debug programs in more effective way.

**Comment by RossComputerGuy** [comment]
Which functions are missing? Can we get a list in a reply and not a zip file please, that'll make it easier.

**Comment by twaik** [comment]
zip file contains patch adding versionsort function implementation to kernel.c patch, but I think that it should be implemented in libandroid-support. Also termux-packages's NDK misses spawn.h and corresponding functions (original NDK has it, but with annotation INTRODUCED_IN(28)).

**Comment by twaik** [comment]
zip file contains build.sh + kernel.c patch

**Comment by twaik** [comment]
Ninja also needs posix_spawn so I decided to make a separate package.
[libandroid-spawn.zip](https://github.com/termux/termux-packages/files/3539202/libandroid-spawn.zip)

**Comment by twaik** [comment]
uftrace builds, but fails with error:
```
$ uftrace tool
child terminated by signal: 11: Segmentation fault
WARN: cannot open record data: /data/data/com.termux/files/usr/tmp/uftrace-live-enMZF9: m
```
If you want to test any tool, you need to build it with `-finstrument-functions` in `$CFLAGS` and `$LDFLAGS`. If program has no `__cyg_profile_func_{enter|leave}` you need to link it with libfakeprofiler.so. 
[uftrace.zip](https://github.com/termux/termux-packages/files/3539299/uftrace.zip)

For now I am to tired to fix this...

**Comment by twaik** [comment]
I used i3's posix shm implementation but error is still the same. Also I fixed getting strerror. For now error message is: 
```
$ uftrace ./s
child terminated by signal: 11: Segmentation fault
WARN: cannot open record data: /data/data/com.termux/files/usr/tmp/uftrace-live-eikpCY: No data available
```
I think it answers ENODATA because @PREFIX@/tmp/uftrace-live-%random% is not created by child.

[uftrace.zip](https://github.com/termux/termux-packages/files/3541096/uftrace.zip)

**Comment by twaik** [comment]
Looks like libfakeprofiler.so is unneded, there is libmcount-nop.so.

**Comment by xeffyr** [comment]
Package is now available. Android 7 only since there problems on android-5.

> Looks like libfakeprofiler.so is unneded, there is libmcount-nop.so.

So libfakeprofiler can be removed in favor of libmcount-nop.so ?

**Comment by twaik** [comment]
> So libfakeprofiler can be removed in favor of libmcount-nop.so ?

Yes.

**Comment by twaik** [comment]
I have a gdb trace of uftrace's child, but I do not understand the reason it behaves this way. Looks like the reason is termux-packages internals.
```
$ gdb -q --args uftrace record tool
Reading symbols from uftrace...
(No debugging symbols found in uftrace)
(gdb) b main
Function "main" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (main) pending.
(gdb) set follow-fork-mode child
(gdb) run
Starting program: /data/data/com.termux/files/usr/bin/uftrace record tool
[Attaching after process 9824 fork to child process 9828]
[New inferior 2 (process 9828)]
[Detaching after fork from parent process 9824]
[Inferior 1 (process 9824) detached]
process 9828 is executing new program: /data/data/com.termux/files/home/tool

Thread 2.1 "tool" received signal SIGSEGV, Segmentation fault.
[Switching to process 9828]
0xb6f355f8 in __dl__ZN12LinkerLogger10ResetStateEv () from /system/bin/linker
(gdb) step
Single stepping until exit from function __dl__ZN12LinkerLogger10ResetStateEv,
which has no line number information.

Thread 2.1 "tool" received signal SIGSEGV, Segmentation fault.
0xb6f353e8 in __dl__ZZL29__linker_init_post_relocationR19KernelArgumentBlockEN3$_28__invokeEv () from /system/bin/linker
(gdb) bt
#0  0xb6f353e8 in __dl__ZZL29__linker_init_post_relocationR19KernelArgumentBlockEN3$_28__invokeEv () from /system/bin/linker
#1  0xb6f3dc3c in __dl__ZL24debuggerd_signal_handleriP7siginfoPv () from /system/bin/linker
#2  <signal handler called>
#3  0xb6f355f8 in __dl__ZN12LinkerLogger10ResetStateEv () from /system/bin/linker
#4  0xb6f2847a in __dl__Z10dlsym_implPvPKcS1_PKv () from /system/bin/linker
#5  0xb6d4e0c6 in dlsym () from /system/lib/libdl.so
#6  0xb6cd5fd4 in ?? () from /data/data/com.termux/files/usr/lib/libmcount-fast.so
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb) quit
A debugging session is active.

	Inferior 2 [process 9828] will be killed.

Quit anyway? (y or n) y
child terminated by signal: 9: Killed
```

**Comment by xeffyr** [comment]
>  Looks like the reason is termux-packages internals.

Looks like the reason is **android linker**.

**Comment by twaik** [comment]
Termux resets some environment variables, maybe that is a reason.
Is there a way to fix it?

**Comment by xeffyr** [comment]
> Termux resets some environment variables, maybe that is a reason.

How you detected that cause is resetted environment variables ?

**Comment by xeffyr** [comment]
What I see from log is only problems with relocation.

**Comment by twaik** [comment]
I have 8.1.0. https://android.googlesource.com/platform/bionic/+/refs/tags/android-8.1.0_r67/linker/linker_main.cpp#211 . But that does not explain why it is segfaulting.

**Comment by xeffyr** [comment]
Seems like /data/data/com.termux/files/usr/lib/libmcount-fast.so hooks some functions in Android's linker.

**Comment by twaik** [comment]
Looks like libmcount-fast.so hooks dlopen. Can that be a problem?

**Comment by xeffyr** [comment]
With strace I noticed that something tries access /dev/shm.

**Comment by xeffyr** [comment]
```
munmap(0x7ac5636000, 131072)            = 0
openat(AT_FDCWD, "/dev/shm/", O_RDONLY|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0x21656361600948} ---
rt_sigaction(SIGSEGV, NULL, {sa_handler=0x7ac58d43c8, sa_mask=~[KILL STOP RTMIN RT_1 RT_2 RT_3], sa_flags=SA_ONSTACK|SA_RESTART|SA_SIGINFO}, 8) = 0
prctl(PR_GET_NO_NEW_PRIVS, 0, 0, 0, 0)  = 0
prctl(PR_GET_NAME, "uftrace")           = 0
```

**Comment by xeffyr** [comment]
The backtrace in https://github.com/termux/termux-packages/issues/4216#issuecomment-527412579 can be wrong. I didn't able to get same.

If I won't do
```
Make breakpoint pending on future shared library load? (y or [n]) y
```
it finishes under gdb without segfault.

**Comment by twaik** [comment]
Yes, looks like cmds/record.c still has /dev/shm reference. I'll get rid of it.

**Comment by xeffyr** [comment]
Done in https://github.com/termux/termux-packages/commit/c29aed6e7bfc30eff9304183e29bd36a97dae4d4

**Comment by twaik** [comment]
Ok. I've rebuild it with "-rdynamic -g" and with #define static __attribute__((visibility("default"))) . I have a new backtrace. 
```
Thread 2.1 "tool" received signal SIGSEGV, Segmentation fault.
[Switching to process 476]
0xb6f355f8 in __dl__ZN12LinkerLogger10ResetStateEv () from /system/bin/linker
(gdb) bt
#0  0xb6f355f8 in __dl__ZN12LinkerLogger10ResetStateEv () from /system/bin/linker
#1  0xb6f2847a in __dl__Z10dlsym_implPvPKcS1_PKv () from /system/bin/linker
#2  0xb6ce60c6 in dlsym () from /system/lib/libdl.so
#3  0xb6d4d220 in mcount_hook_functions () from /data/data/com.termux/files/usr/lib/libmcount-fast.so
#4  0xb6d4c1ea in mcount_startup () from /data/data/com.termux/files/usr/lib/libmcount-fast.so
#5  0xb6f385de in __dl__ZL13call_functionPKcPFviPPcS2_ES0_ () from /system/bin/linker
#6  0xb6f3833c in __dl__ZL10call_arrayIPFviPPcS1_EEvPKcPT_jbS5_ () from /system/bin/linker
#7  0xb6f384e4 in __dl__ZN6soinfo17call_constructorsEv () from /system/bin/linker
#8  0xb6f383da in __dl__ZN6soinfo17call_constructorsEv () from /system/bin/linker
#9  0xb6f3507a in __dl___linker_init () from /system/bin/linker
#10 0xb6f39f70 in __dl__start () from /system/bin/linker
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

I tried to check which dlsym in mcount_hook_functions fails, but looks like printf does not work here.

**Comment by xeffyr** [comment]
Segmentation fault should be fixed in https://github.com/termux/termux-packages/commit/877571c0ee1b27a28f8acb4340c34f63ff5cfd17.

Now remaining problem is that no uftrace data generated for ARM.
